settings/src/components/password.js

```diff
@@ -1,21 +1,72 @@
+/**
+ * External dependencies
+ */
+import { pick } from 'lodash';
+// window.zxcvbn is also used, but can't be declared here because it's loaded asyncronously.
+
 /**
  * WordPress dependencies
  */
-import { Button, TextControl } from '@wordpress/components';
+import { Button, Flex, Notice, TextControl }     from '@wordpress/components';
+import { useCallback, useEffect, useState }      from '@wordpress/element';
+import { Icon, cancelCircleFilled, check, seen } from '@wordpress/icons';
+
+import {} from '@wordpress/util'; // Enqueue via asset.php.
+const { post } = wp.ajax;
 
 /**
  * Render the Password setting.
  */
-export default function Password() {
-	// todo
-	// use same generator as core, do server side?
-	// strength meter
-	// warn when weak
-	// automatically generate placeholder?
-	// how to get empty value so placeholder shows up?
+export default function Password( { userRecord, userRequires2fa } ) {
+	const [ passwordStrong, setPasswordStrong ] = useState( false );
+	const [ saved, setSaved ]                   = useState( false );
+	const [ generating, setGenerating ]         = useState( false );
+	const [ inputType, setInputType ]           = useState( 'password' );
+
+	useEffect( () => {
+		if ( ! userRecord.editedRecord.password ) {
+			return;
+		}
+
+		setSaved( false );
+		setPasswordStrong( isPasswordStrong( userRecord.editedRecord.password, userRecord.record, userRequires2fa ) );
+	}, [ userRecord.editedRecord.password ] );
+
+	/**
+	 * @todo When starting from a fresh page load, theres a flash of the red "too easy" notice before it shows
+	 * the success notice. It happens between the time when `edit()` is called and the `useEffect` callback fires
+	 * to update the `passwordStrong` state. Is there a way to tell React to wait until both are done to re-render?
+	 * Or maybe the condition that renders the notice can include something like `hasResolved`?
+	 */
+	const generatePassword = useCallback( async () => {
+		setGenerating( true );
+		const password = await post( 'generate-password' );
```

Comment:
Also a good idea 👍🏻

Reply:
...aaaaaand that turned into a rabbit hole 😣

Even after throwing out the answers on that SO thread that don't use `crypto.getRandomValues()`, it's still not really clear what the proper way to do it is.

JS doesn't provide a native way to uniformly get a cryptographically-secure random number _within a range_, and doing it manually seems [complicated](https://stackoverflow.com/questions/41437492/how-to-use-window-crypto-getrandomvalues-to-get-random-values-in-a-specific-rang) and [error-prone](https://dimitri.xyz/random-ints-from-random-bits/).

I searched for info from a security expert about using `getRandomValues()` to generate a password, but couldn't find any. Given how nuanced the topic is and how many different methods there are, I'm leery to rely on an SO answer.

Maybe I'm being too paranoid, though, and [mwag's answer](https://stackoverflow.com/a/43020177/450127r) is Good Enough? I've cleaned it up in 0b76f5b to make it easier to read and try out. The part I'm not sure about is if it's really a uniform distribution like the author claims, or if using the modulus creates a bias.

I might take a closer look later, but I think I've probably spent too much time on it for now, and am leaning towards sticking w/ the AJAX request.

Reply:
Using `%` does introduce a bias, but given the small array size of characters, it's not going to be _that bad_ imho.

I certainly wouldn't rely upon it for key generation, but for a psuedo-random password it seems _good enough_.

Realistically, using random character passwords is not ideal regardless, as most people are more likely to use their own password OR they use a password manager which would insert a much better random password, rather than what we should see as a **strong password**. We don't want people using the random generated passwords, and NOT using a password manager..

I took a look what Automattic uses on Calypso, and this is what I came up with: https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/generate-password/index.js

Is there a 3rd-party library we can just import and rely upon for offering stronger passwords perhaps? Maybe that generates it based on a wordlist? 

Sticking with the XHR for now is fine by me, if it's easier, it's probably worthwhile looking at alternatives long-term.

Reply:
Looks like PHP's `random_int()` does something similar to 0b76f5bac57608aa40b22d1d99c6ee0e84d91d7f - It just keeps generating numbers until something falls within the range: https://github.com/php/php-src/blob/master/ext/random/random.c#L623-L665 although it has some optimization as powers of two aren't biased apparently.

Reply:
Yeah, I agree it's ideal if folks use a password manager or their browser to generate it, since those aret in a better position to ensure randomness. In a few cases the entropy may be lower on paper, like when Firefox generates passwords that are only 15 characters, but that's probably maybe not a big deal 🤷🏻‍♂️ 

For a pass_phrase_ library, [niceware](https://www.npmjs.com/package/niceware) seems like the best from a security perspective, but it's completely unusable for us since their dictionary contains vulgar words. Using the [EFF large wordlist](https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases) would be a requirement IMO. [eff-diceware-passphrase](https://www.npmjs.com/package/eff-diceware-passphrase) also looks good, but is written for Node, and porting it along with the related dependencies wouldn't be worth it. I couldn't anything else that I'd trust, unfortunately 🙁

For a pass_word_ library, [password-generator](https://www.npmjs.com/package/password-generator) looks pretty good based on usage, but I didn't see anything using `crypto.getRandomValues()` written by someone that I can tell is an expert.

Reply:
...oh, wait, looking at the code there, it looks like if we just use a `UInt8Array` instead of a `UInt32Array` then the random numbers are bounded to `0-255` 🤦🏻‍♂️ 

So then it's performant to use rejection sampling to avoid bias, which is what that library does. I went ahead and did that in 5241afe